uspace/app/bdsh/cmds/modules/ls/ls.c

```diff
@@ -89,14 +97,45 @@ static unsigned int ls_start(ls_job_t *ls)
  */
 static void ls_print(struct dir_elem_t *de)
 {
-	if (de->s.is_file)
-		printf("%-40s\t%llu\n", de->name, (long long) de->s.size);
-	else if (de->s.is_directory)
-		printf("%-40s\t<dir>\n", de->name);
+	int width = 13;
+
+	if (de->s.is_file) {
+		if (ls.well_formatted) {
+			cap_spec_t cap;
+
+			cap_from_blocks(1, de->s.size, &cap);
+			cap_simplify(&cap);
+
+			char *rptr;
+			if (cap_format(&cap, &rptr) == EOK) {
+				char bytes[9], suffix[3];
+				sscanf(rptr, "%s %s", bytes, suffix);
+				free(rptr);
+
+				printf("%-40s\t%*s %2s\n", de->name, width - 3, bytes, suffix);
+				
+				//if there is a failure with cap_format we simply print out an unformatted size
```

Comment:
cap_format will only fail if we run out of memory. In general, OOM errors should be handled as follows: print an error message "Out of memory.", stop what you are doing and return to the closest interactive loop. In this case you should print "Out of memory.", stop listing the directory and return a non-zero exit code from ls.